db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -1087,21 +1087,60 @@ Status DBImpl::CompactRangeInternal(const CompactRangeOptions& options,
     {
       SuperVersion* super_version = cfd->GetReferencedSuperVersion(this);
       Version* current_version = super_version->current;
+
+      // Might need to query the partitioner
+      SstPartitionerFactory* partitioner_factory =
+          current_version->cfd()->ioptions()->sst_partitioner_factory.get();
+      std::unique_ptr<SstPartitioner> partitioner;
+      if (partitioner_factory && begin != nullptr && end != nullptr) {
+        SstPartitioner::Context context;
+        context.is_full_compaction = false;
+        context.is_manual_compaction = true;
+        context.output_level = /*unknown*/ -1;
+        // Small lies aobut compaction range
+        context.smallest_user_key = *begin;
+        context.largest_user_key = *end;
+        partitioner = partitioner_factory->CreatePartitioner(context);
+      }
+
       ReadOptions ro;
       ro.total_order_seek = true;
       bool overlap;
       for (int level = 0;
            level < current_version->storage_info()->num_non_empty_levels();
            level++) {
         overlap = true;
+
+        bool check_overlap_within_file = false;
         if (begin != nullptr && end != nullptr) {
+          // Typically checking overlap within files in this case
+          check_overlap_within_file = true;
+          if (partitioner) {
+            // Especially if the partitioner is new, the manual compaction
+            // might be used to enforce the partitioning. So if there is a
+            // partitioner and the compaction range crosses a partitioning
+            // boundary, then file-level overlap info is sufficient for
+            // deciding whether to consider the level as overlapping.
+            // (Proof idea: we want to consider overlapping if some file
+            // crosses a boundary in range; otherwise, if there's a boundary
+            // in range and some actual entries overlap in range, then a file
+            // smallest or largest must be in range.
```

Comment:
Is this trying to prove "if a file crosses a partition boundary within the compaction range, then the file's smallest or largest must be in range (i.e, file-level overlap checks is enough to find this file)??

Reply:
Two things I am interested in seeing some sort of proof are:
(1) File-level overlap checking is sufficient to include all files crossing the partition boundary in the compact range (i.e, sufficient proof)
(2) File-level overlap checking only includes (a) crossing-partition-boundary files and (b) non-crossing-partition-boundary files that would have been included in within-file overlap checking. (i.e, regression test, that is we don't include useless files compared with the previous check_overlap_within_file implementation)

Both statements are true IMO:
(1) If a file crosses the partition boundary, then the file contains a key k1 in 1st partition and k2 in 2nd partition. Since SST file is sorted by keys within, the smallest key <= k1 in 1st partition and the largest key >= k2 in 2nd partition. Since the partition boundary is in the compaction range, the compaction range will have file-level overlap with the file. That is file-level overlap checking is sufficient to include this file.

(2) If a file file-level-overlaps with the compaction and assume it would NOT have been included under within-file overlap checking, then the file does not contains any keys in compaction range. But the file is file-overlapping with the compaction range, then the file's smallest key will be smaller than the compaction range and the largest key will be greater than the compaction range. Therefore the file range will contains the compaction range and will cross the partition boundary contained in the compaction range.

Reply:
Above are the verbose version of my proof, to make sure I understand the proof idea here. The proof idea above is very very precise that I am not really sure which thing it is proving. Will appreciate more explanation above (not necessarily as verbose as mine).

Reply:
You've got the right idea. I'll rephrase everything and hopefully it's simple and concise.